SpatialGDK/Source/SpatialGDKEditor/Private/SpatialGDKEditorPackageAssembly.cpp

```diff
@@ -0,0 +1,209 @@
+	FString Project = FPaths::ConvertRelativePathToFull(FPaths::GetProjectFilePath());
+	FString Args = FString::Printf(TEXT("%s %s %s %s"), *ProjectName, *Platform, *Configuration, *Project, *AdditionalArgs);
+	PackageAssemblyTask = MakeShareable(new FMonitoredProcess(SpatialBuildExe, Args, WorkingDir, true));
+	PackageAssemblyTask->OnCompleted().BindRaw(this, &FSpatialGDKPackageAssembly::OnTaskCompleted);
+	PackageAssemblyTask->OnOutput().BindRaw(this, &FSpatialGDKPackageAssembly::OnTaskOutput);
+	PackageAssemblyTask->OnCanceled().BindRaw(this, &FSpatialGDKPackageAssembly::OnTaskCanceled);
+	PackageAssemblyTask->Launch();
+	FString NotificationMessage = FString::Printf(TEXT("Building %s Assembly"), *ProjectName);
+}
+
+void FSpatialGDKPackageAssembly::UploadAssembly(const FString &AssemblyName, bool Force)
+{
+	const USpatialGDKSettings* SpatialGDKSettings = GetDefault<USpatialGDKSettings>();
+	FString WorkingDir = FPaths::ConvertRelativePathToFull((FPaths::IsProjectFilePathSet() ? FPaths::GetPath(FPaths::GetProjectFilePath()) : FPaths::RootDir() / FApp::GetProjectName()) / TEXT("..") / TEXT("spatial"));
+	FString Spatial(TEXT("spatial"));
+	FString Args = FString::Printf(TEXT("cloud upload %s %s %s --no_animation"), *AssemblyName, Force ? TEXT("--force") : TEXT(""), SpatialGDKSettings->IsRunningInChina() ? TEXT("--environment=cn-production") : TEXT(""));
+	PackageAssemblyTask = MakeShareable(new FMonitoredProcess(Spatial, Args, WorkingDir, true));
+	PackageAssemblyTask->OnCompleted().BindRaw(this, &FSpatialGDKPackageAssembly::OnTaskCompleted);
+	PackageAssemblyTask->OnOutput().BindRaw(this, &FSpatialGDKPackageAssembly::OnTaskOutput);
+	PackageAssemblyTask->OnCanceled().BindRaw(this, &FSpatialGDKPackageAssembly::OnTaskCanceled);
+	PackageAssemblyTask->Launch();
+	FString NotificationMessage = FString::Printf(TEXT("Uploading Assembly to Project: %s"), *FSpatialGDKServicesModule::GetProjectName());
+	if (AssemblyDetailsPtr)
+	{
+		AssemblyDetailsPtr.Reset(nullptr);
+	}
+}
+
+void FSpatialGDKPackageAssembly::BuildAllAndUpload(const FString& AssemblyName, const FString& WindowsPlatform, const FString& Configuration, const FString& AdditionalArgs, bool Force)
+{
+	if (!AssemblyDetailsPtr && CurrentAssemblyTarget == EPackageAssemblyTarget::NONE)
+	{
+		AssemblyDetailsPtr.Reset(new AssemblyDetails(AssemblyName, Configuration, Force));
+		CurrentAssemblyTarget = EPackageAssemblyTarget::BUILD_CLIENT;
+		BuildAssembly(FApp::GetProjectName(), WindowsPlatform, Configuration, AdditionalArgs);
+
+		AsyncTask(ENamedThreads::GameThread, [this]() { this->ShowTaskStartedNotification(TEXT("Building Assembly"));	});
+	}
+}
+
+bool FSpatialGDKPackageAssembly::CanBuild() const
+{
+	return CurrentAssemblyTarget == EPackageAssemblyTarget::NONE;
+}
+
+void FSpatialGDKPackageAssembly::OnTaskCompleted(int32 Result)
```

Comment:
is this for the one button for building out client, worker, sim player and assembly?
do we also have one where you can e.g. only build the client?

Reply:
What's the use case? From the Cloud Deploy dialog, my understanding is we want to build and upload in order to successfully launch the deployment? What value would their be in only building a specific client? And why would it belong on this particular dialog?